### DISCUSSION TITLE
Sensor to detect when camera in IR mode (ie IsBlackAndWhite)?

### DISCUSSION THREAD

**Discussion by leadZERO** [blakeblackshear/frigate#6851] [Ideas]
One thing I've struggled with over the years of doing home automation is finding a good way to detect when lights would be useful. I have a weather station that has a solar radiance sensor, I have indoor light sensors. I've tried various combinations but so far haven't come up with a great solution. 

In the meantime I use the ol' "sunrise/sunset" kind of timing offset. The problem with this is it doesn't work well in some edge cases, cloudiness and/or certain times of year being particularly problematic.

One thing I've noticed is that my cameras are really good at detecting "too little" light. They use it, of course, to switch to IR mode. However, I've found this to be a great proxy for when things like outdoor lights and/or indoor lights are needed. And unlike basing it off sunrise/sunset offset times, it works directly off what I want to measure... "How hard is it to see without supplemental light?"

Would it be possible to add this kind of sensor into Frigate (and the Home assistant devices)? I imagine the amount of processing would be incredibly small, even running a calculation on a frame every other second would probably be sufficient. I imagine this kind of calculation would be simple with the video frame data.

Would anyone else find this useful? Alternatively, has anyone else found a great way to measure light levels low enough to use as a signal when indoor/outdoor lights are needed?

**Comment by NickM-27** [comment]
I would suggest if you use it to turn lights on it may not work the way you think. Imagine the case where a camera switches to IR mode and then an automation turns lights on so then the camera switches back to color mode and then the automation turns off the lights... rinse and repeat.

Anyway, outside of that, I'm sure this will be possible.

**Reply by leadZERO** [reply]
Definitely! And that kind of logic I would imagine would be added in the automation layer. What I'm looking for is just a simple dumb sensor that detects when the camera is in IR mode and when it isn't. I've already looked in the data the camera(s) can send, but isn't there in my particular case. However, I think this kind of functionality would be quite useful for many different people. And the calculation is cheap enough that adding it from frame data that Frigate is already digesting is simpler than a per-model kind of status of the IR LEDs.

**Comment by moandcompany** [comment]
> One thing I've struggled with over the years of doing home automation is finding a good way to detect when lights would be useful. I have a weather station that has a solar radiance sensor, I have indoor light sensors. I've tried various combinations but so far haven't come up with a great solution.
> 
> In the meantime I use the ol' "sunrise/sunset" kind of timing offset. The problem with this is it doesn't work well in some edge cases, cloudiness and/or certain times of year being particularly problematic.
> 
> One thing I've noticed is that my cameras are really good at detecting "too little" light. They use it, of course, to switch to IR mode. However, I've found this to be a great proxy for when things like outdoor lights and/or indoor lights are needed. And unlike basing it off sunrise/sunset offset times, it works directly off what I want to measure... "How hard is it to see without supplemental light?"
> 
> Would it be possible to add this kind of sensor into Frigate (and the Home assistant devices)? I imagine the amount of processing would be incredibly small, even running a calculation on a frame every other second would probably be sufficient. I imagine this kind of calculation would be simple with the video frame data.
> 
> Would anyone else find this useful? Alternatively, has anyone else found a great way to measure light levels low enough to use as a signal when indoor/outdoor lights are needed?

You can also determine this signal by using the Frigate or camera's built-in snapshot function and calculating luminance levels across color channels for the snapshot image.

**Reply by x-real-ip** [reply]
Hi, I made is-it-dark, a small API that checks if an image is dark or bright. It also returns the calculated brightness value and detects if the camera is in infrared mode. It’s useful for automations or analyzing camera images. The service can run containerised.

https://github.com/x-real-ip/is-it-dark

**Comment by OnFreund** [comment]
I use a custom image processing integration in HA for that. What I do is sample 10 pixels, and if R == G == B for all of them, I assume the camera is in IR mode.
